agent/php_execute.c

```diff
@@ -919,40 +919,97 @@ static void nr_php_execute_file(const zend_op_array* op_array,
- * being called is a __call() magic method (in which case a trampoline is
- * created and destroyed). We increment the reference counts on the scope and
- * function strings and keep pointers to them in this structure, then release
- * them once we've named the trace node and/or metric (if required).
- *
- * In PHP 5, execute_data->op_array may be set to NULL if we make a subsequent
- * user function call in an exec callback (which occurs before we decide
- * whether to create a metric and/or trace node), so we keep a copy of the
- * pointer here. The op array itself won't be destroyed from under us, as it's
- * owned by the request and not the specific function call (unlike the
- * trampoline case above).
+ * Params  : 1. A pointer to a metadata structure.
+ *           2. The zend_execute_data
  *
- * Note that, while op arrays are theoretically reference counted themselves,
- * we cannot take the simple approach of incrementing that reference count due
- * to not all Zend Engine functions using init_op_array() and
- * destroy_op_array(): one example is that PHP 7 trampoline op arrays are
- * simply emalloc() and efree()'d without even setting the reference count.
- * Therefore we have to be more selective in our approach.
+ * Note    : It is the responsibility of the caller to allocate the metadata
+ *           structure. In general, it's expected that this will be a pointer
+ *           to a stack variable.
  */
-typedef struct {
-#ifdef PHP7
-  zend_string* scope;
-  zend_string* function;
+static void nr_php_execute_metadata_add_code_level_metrics(
+    nr_php_execute_metadata_t* metadata,
+    NR_EXECUTE_PROTO) {
+#if ZEND_MODULE_API_NO < ZEND_7_0_X_API_NO /* PHP7+ */
+  (void)metadata;
+  NR_UNUSED_SPECIALFN;
+  return;
 #else
-  zend_op_array* op_array;
+  const char* filepath = NULL;
+  const char* namespace = NULL;
+  const char* function = NULL;
+  uint32_t lineno = 1;
+
+  if (NULL == metadata) {
+    return;
+  }
+
+  if (NULL == execute_data) {
+    return;
+  }
+
+  metadata->function_name = NULL;
+  metadata->function_filepath = NULL;
+  metadata->function_namespace = NULL;
+
+  /*
+   * Check if code level metrics are enabled in the ini.
+   * If they aren't, exit and don't update CLM.
+   */
+  if (!NRINI(code_level_metrics_enabled)) {
+    return;
+  }
+  /*
+   * At a minimum, at least one of the following attribute combinations MUST be
+   * implemented in order for customers to be able to accurately identify their
+   * instrumented functions:
+   *  - code.filepath AND code.function
+   *  - code.namespace AND code.function
+   *
+   * If we don't have the minimum requirements, exit and don't add any
+   * attributes.
+   */
+
+  filepath = nr_php_zend_execute_data_filename(execute_data);
+  namespace = nr_php_zend_execute_data_scope_name(execute_data);
+  function = nr_php_zend_execute_data_function_name(execute_data);
+  lineno = nr_php_zend_execute_data_lineno(execute_data);
+
```

Comment:
check all the string lengths with NRSAFELEN.  If any exceed 255 exit without adding here.

Reply:
implemented in 125fa65bc36a89858fcadcff05d5aacf447ee533